### PR TITLE
Use the new Fenix components model to move bugs out of Fenix:General

### LINKED ISSUE #2398: Use the new Fenix components model to move bugs out of Fenix:General
Once https://github.com/mozilla/bugbug/issues/4172 is resolved, we should use the new model to move bugs from the `Fenix:General` component to other Fenix components.

The model should be integrated as part of the [component](https://github.com/mozilla/bugbot/blob/master/bugbot/rules/component.py) rule and should be utilized in the following two scenarios:
- When the component model classifies a bug into `Fenix:General` component.
- When the component model is not moving a bug already in the `Fenix:General` to another Fenix component.

### PR DESCRIPTION
**Pull request by benjaminmah**
Resolves #2398.

Intended to use the Fenix-specific component classifier model from [mozilla/bugbug#4172](https://github.com/mozilla/bugbug/issues/4172) to move bugs that already exist in `Fenix:General` or have been classified in `Fenix:General` by the general `component` model to the other Fenix components.

**NOTE:** Dependent on [mozilla/bugbug#4172](https://github.com/mozilla/bugbug/issues/4172) to be merged first to use the Fenix-specific component classifier model.

## Checklist

- [ ] Type annotations added to new functions
- [ ] Docs added to functions touched in main classes
- [x] Dry-run produced the expected results
- [x] The [`to-be-announced`](https://github.com/mozilla/bugbot/labels/to-be-announced) tag added if this is worth announcing

### CODE REVIEW

File: bugbot/rules/component.py [hunk 1]
```diff
@@ -91,6 +91,9 @@ def get_bugs(self, date="today", bug_ids=[]):
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:91]: I would add the logic to call the Fenix model here to avoid code duplicates.

File: bugbot/rules/component.py [hunk 2]
```diff
@@ -77,6 +77,96 @@ def get_bz_params(self, date):
         }
 
     def get_bugs(self, date="today", bug_ids=[]):
+        def process_bugs(bugs, raw_bugs, results, reclassify_fenix):
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:80]: Could we avoid having a big diff in the PR?

File: bugbot/rules/component.py [hunk 3]
```diff
@@ -89,6 +89,21 @@ def get_bugs(self, date="today", bug_ids=[]):
         # Classify those bugs
         bugs = get_bug_ids_classification("component", bug_ids)
 
+        # Collect bugs classified as Fenix::General
+        fenix_general_bug_ids = [
+            bug_id
+            for bug_id, bug_data in bugs.items()
+            if "class" in bug_data and bug_data["class"] == "Fenix::General"
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:96]: We could simplify this by using `get()`:
```suggestion
            if bug_data.get("class") == "Fenix::General"
```

File: bugbot/rules/component.py [hunk 4]
```diff
@@ -89,6 +89,21 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+        ]
+
+        # Reclassify the Fenix::General bugs using the fenixcomponent model
+        if fenix_general_bug_ids:
+            fenix_general_classification = get_bug_ids_classification(
+                "fenixcomponent", fenix_general_bug_ids
+            )
+            for bug_id in fenix_general_classification:
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:104]: nit: We could use `items()` here to iterate on both the key and the value simultaneously.

File: bugbot/rules/component.py [hunk 5]
```diff
@@ -89,6 +89,21 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+            if bug_data.get("class") == "Fenix::General"
+        ]
+
+        # Reclassify the Fenix::General bugs using the fenixcomponent model
+        if fenix_general_bug_ids:
+            fenix_general_classification = get_bug_ids_classification(
+                "fenixcomponent", fenix_general_bug_ids
+            )
+            for bug_id, data in fenix_general_classification.items():
+                bugs[bug_id] = data
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:105]: The downside with this solution is that if we have high confidence to move a bug to `Fenix::General`, then the Fenix model gives us low confidence for the specific Fenix component, we will end up doing nothing. Ideally, moving the bug to `Fenix::General` is better than nothing.

**Comment by benjaminmah** [comment]
Results from training general component model with all `Fenix` bugs labelled as `Fenix::General`

https://docs.google.com/document/d/11tk2N7A0GxIvA8aWorqt6eSuzKgZQpPG6_8yycH8wmM/edit?usp=sharing

File: bugbot/rules/component.py (hunk 1 above)
Comment (benjaminmah) [reply bugbot/rules/component.py:91]: Done in aa18d50c9a5635fc36a2ee15a0ab4d7b38adb945

File: bugbot/rules/component.py (hunk 3 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Done in 246a34614bbe11b6aa3783164a10ce6336f20d09!

File: bugbot/rules/component.py (hunk 4 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Done in 246a34614bbe11b6aa3783164a10ce6336f20d09!

File: bugbot/rules/component.py (hunk 5 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: In ae5f471484317064592b92a655a8b6c59841fb41, I've added a check to ensure that if the new classification does not meet the threshold, we default to the original classification from the general model.

File: bugbot/rules/component.py [hunk 6]
```diff
@@ -89,6 +89,36 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+            bug_id
+            for bug_id, bug_data in bugs.items()
+            if bug_data.get("class") == "Fenix::General"
+        ]
+
+        # Reclassify the Fenix::General bugs using the fenixcomponent model
+        if fenix_general_bug_ids:
+            fenix_general_classification = get_bug_ids_classification(
+                "fenixcomponent", fenix_general_bug_ids
+            )
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:103]: With this, we use the model to predict the Fenix component for any bug that the component model classifies as `Fenix::General` even if they have very low confidence. Such bugs are usually filtered out of the training set for the Fenix-only model. My concern here is that we are exposing the model to types of bugs that were not trained on.

File: bugbot/rules/component.py [hunk 7]
```diff
@@ -89,6 +89,36 @@ def get_bugs(self, date="today", bug_ids=[]):
         # Classify those bugs
         bugs = get_bug_ids_classification("component", bug_ids)
 
+        # Collect bugs classified as Fenix::General
+        fenix_general_bug_ids = [
+            bug_id
+            for bug_id, bug_data in bugs.items()
+            if bug_data.get("class") == "Fenix::General"
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:96]: I think the model will output `Fenix` instead of `Fenix::General`.

Comment (benjaminmah) [reply bugbot/rules/component.py]: Fixed in 1af5b0ab6b9c86c82c39142aa33404447e673e9e.

File: bugbot/rules/component.py (hunk 6 above)
Comment (benjaminmah) [reply bugbot/rules/component.py:103]: I agree that there could be cases where the Fenix-specific model could encounter bug types that it has not been trained on. What do you think of increasing the component confidence threshold for the Fenix-specific model? If it does not meet the threshold it would just default to the general component model classification (i.e. `Fenix::General`).

Comment (suhaibmujahid) [reply bugbot/rules/component.py:103]: Yes, we should use the threshold that gave us good results on the testing dataset.

Comment (benjaminmah) [reply bugbot/rules/component.py:103]: I've introduced a new confidence threshold for Fenix-specific components in `rules.json` in db0adb583d5e336d7dd41cf945d37ab1c606dea6. It's currently set to 0.6 (from [here](https://docs.google.com/spreadsheets/d/1ocDf-sTuk8AFqUDBYtYMUtR7uwJ0Nk5_8k7EO9DsXCI/edit?usp=sharing)) but can be changed accordingly.

Comment (suhaibmujahid) [reply bugbot/rules/component.py:103]: I see that "No CT" gives us higher recall without losing on the precision compared to "0.6". What do you think?

**Comment by benjaminmah** [comment]
@suhaibmujahid 

Spreadsheet on reclassified bugs that were originally classified as `Fenix::General`:

https://docs.google.com/spreadsheets/d/1MiAhKmdUx_TER7NOPQ7TAzFTfKJrYGsrlAmDe3p13EQ/edit?usp=sharing

File: bugbot/rules/component.py [hunk 8]
```diff
@@ -89,6 +89,55 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+            and bug_data["prob"][bug_data["index"]]
+            >= self.get_config("general_confidence_threshold")
+        ]
+
+        def get_confidence_threshold(
+            bug_data, general_confidence_threshold, confidence_threshold
+        ):
+            if bug_data["class"] == "General":
+                return general_confidence_threshold
+            return confidence_threshold
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:106]: We could simplify this by changing the function to check whether it passes the threshold or not.

Selecting a good name for the function will be important to improve the readability.

File: bugbot/rules/component.py [hunk 9]
```diff
@@ -89,6 +89,55 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+        # Collection bugs that were originally Fenix::General but reclassified to another product with low confidence
+        originally_fenix_general_bug_ids = [
+            bug_id
+            for bug_id, bug_data in bugs.items()
+            if raw_bugs[bug_id]["product"] == "Fenix"
+            and raw_bugs[bug_id]["component"] == "General"
+            and bug_data["prob"][bug_data["index"]]
+            <= get_confidence_threshold(
+                bug_data,
+                self.get_config("general_confidence_threshold"),
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:117]: We could store all thresholds in variables at the top and reuse the variables instead of repeatedly calling `self.get_config()`.

File: bugbot/rules/component.py [hunk 10]
```diff
@@ -89,6 +89,55 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+                self.get_config("confidence_threshold"),
+            )
+        ]
+
+        fenix_general_bug_ids.extend(originally_fenix_general_bug_ids)
+        fenix_general_bug_ids = set(fenix_general_bug_ids)
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:123]: We could have it as set from the beginning instead of converting it to set later.

File: bugbot/rules/component.py [hunk 11]
```diff
@@ -106,15 +155,17 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+                "conflated_components_mapping", {}
+            )
 
             # Skip product-only suggestions that are not useful.
             if "::" not in suggestion and bug["product"] == suggestion:
                 continue
 
-            suggestion = conflated_components_mapping.get(suggestion, suggestion)
+            if "Fenix" not in suggestion:
+                suggestion = conflated_components_mapping.get(suggestion, suggestion)
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:168]: Does `conflated_components_mapping` have `Fenix`?

File: bugbot/rules/component.py (hunk 8 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Added a new function called `meets_threshold()` that checks if a classification meets the specified threshold: bc68d06132884591efb2653977567238bd0ea681

File: bugbot/rules/component.py (hunk 9 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Stored all the thresholds as variables in the initialization: bc68d06132884591efb2653977567238bd0ea681

The rest of `get_bugs()` still uses `self.get_config()` but it may be out of this PR's scope. WDYT?

File: bugbot/rules/component.py (hunk 10 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Done in bc68d06132884591efb2653977567238bd0ea681!

File: bugbot/rules/component.py (hunk 11 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: I originally had this check because I encountered errors about `Fenix` not being in `conflated_components_mapping()` but it seems to run fine now! I've removed this conditional here: bc68d06132884591efb2653977567238bd0ea681

File: bugbot/rules/component.py (hunk 8 above)
Comment (suhaibmujahid) [reply bugbot/rules/component.py]: I think it will be cleaner to select the threshold inside `meets_threshold()`. All that we need to pass is `bug_data`.

File: bugbot/rules/component.py (hunk 9 above)
Comment (suhaibmujahid) [reply bugbot/rules/component.py]: Yes, we could do it later in a separate PR. Feel free to file a follow-up issue.

File: bugbot/rules/component.py [hunk 12]
```diff
@@ -76,6 +81,9 @@ def get_bz_params(self, date):
             "f9": "CP",
         }
 
+    def meets_threshold(self, bug_data, threshold):
+        return bug_data["prob"][bug_data["index"]] >= threshold
+
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:86]: We might have it as a local function instead. The function should also select the threshold. We could add docstrings to clarify the scope if necessary.

File: bugbot/rules/component.py (hunk 8 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Done here: ff5b63363ce48bf29f2ba730efd91ff7de42a0d5

File: bugbot/rules/component.py (hunk 12 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Done here: ff5b63363ce48bf29f2ba730efd91ff7de42a0d5

File: bugbot/rules/component.py (hunk 9 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: I'll be sure to file an issue once this PR is merged!

Comment (suhaibmujahid) [reply bugbot/rules/component.py]: You could file the issue before merging the PR if you want.

Comment (benjaminmah) [reply bugbot/rules/component.py]: Done here https://github.com/mozilla/bugbot/issues/2439

File: bugbot/rules/component.py [hunk 13]
```diff
@@ -77,6 +82,16 @@ def get_bz_params(self, date):
...
     def get_bugs(self, date="today", bug_ids=[]):
+        def select_threshold(bug_data):
+            if bug_data["class"] == "Fenix" or bug_data["class"] == "General":
+                return self.general_confidence_threshold
+            else:
+                return self.component_confidence_threshold
+
+        def meets_threshold(bug_data):
+            threshold = select_threshold(bug_data)
+            return bug_data["prob"][bug_data["index"]] >= threshold
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:93]: Do we really need the logic here to be in two functions?

Comment (benjaminmah) [reply bugbot/rules/component.py]: You're right, it's probably more clear to have it one function: 805066fecf101b890c0aa47115e8b339e8ccac1e

File: bugbot/rules/component.py [hunk 14]
```diff
@@ -77,6 +82,13 @@ def get_bz_params(self, date):
         }
 
     def get_bugs(self, date="today", bug_ids=[]):
+        def meets_threshold(bug_data):
+            if bug_data["class"] == "Fenix" or bug_data["class"] == "General":
+                threshold = self.general_confidence_threshold
+            else:
+                threshold = self.component_confidence_threshold
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:89]: Nit:
```suggestion
            threshold = (
                self.general_confidence_threshold
                if bug_data["class"] == "Fenix" or bug_data["class"] == "General"
                else self.component_confidence_threshold
            )
```

File: bugbot/rules/component.py [hunk 15]
```diff
@@ -89,6 +101,38 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+        fenix_general_bug_ids.update(originally_fenix_general_bug_ids)
+
+        # Reclassify the Fenix::General bugs using the fenixcomponent model
+        if fenix_general_bug_ids:
+            fenix_general_classification = get_bug_ids_classification(
+                "fenixcomponent", fenix_general_bug_ids
+            )
+
+            for bug_id, data in fenix_general_classification.items():
+                new_confidence = data["prob"][data["index"]]
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:129]: Why do we call it `new_confidence`?
```suggestion
                confidence = data["prob"][data["index"]]
```

File: bugbot/rules/component.py [hunk 16]
```diff
@@ -89,6 +101,38 @@ def get_bugs(self, date="today", bug_ids=[]):
...
+
+                # Only reclassify if the new confidence meets the Fenix component confidence threshold
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:131]: I recommend not adding comments if they don't add much value. If the code is not clear, I would try to make it clear itself instead of adding a comment; if it is already clear, then a comment to explain "what" would not be necessary. The "why" comments are usually more helpful.

(this applies to all comments in general)

**Review by suhaibmujahid — COMMENTED** [review]
Could you please verify if the result is still similar? Other than that, things looks good to me. Thank you!

File: bugbot/rules/component.py (hunk 16 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Removed here: be25e479c1ce3c954cf85d323a9dcbf6ffc1f4f2

File: bugbot/rules/component.py (hunk 15 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: I agree, `confidence` makes more sense here. Changed here: be25e479c1ce3c954cf85d323a9dcbf6ffc1f4f2

File: bugbot/rules/component.py (hunk 14 above)
Comment (benjaminmah) [reply bugbot/rules/component.py]: Changed here: be25e479c1ce3c954cf85d323a9dcbf6ffc1f4f2

**Comment by benjaminmah** [comment]
> Could you please verify if the result is still similar? Other than that, things looks good to me. Thank you!

Everything looks good on my end!

File: bugbot/rules/component.py [hunk 17]
```diff
@@ -106,9 +147,10 @@ def get_bugs(self, date="today", bug_ids=[]):
             prob = bug_data["prob"]
             index = bug_data["index"]
             suggestion = bug_data["class"]
-            conflated_components_mapping = bug_data["extra_data"][
-                "conflated_components_mapping"
-            ]
+
+            conflated_components_mapping = bug_data["extra_data"].get(
+                "conflated_components_mapping", {}
+            )
```
Comment (suhaibmujahid) [inline bugbot/rules/component.py:153]: Is this change intended? What is the goal?

Comment (benjaminmah) [reply bugbot/rules/component.py:153]: I get a KeyError: `KeyError: 'conflated_components_mapping'` because the reclassified Fenix bugs don't have this field in their `bug_data`.

Comment (suhaibmujahid) [reply bugbot/rules/component.py:153]: I see. Thank you!

**Review by suhaibmujahid — APPROVED** [review]
LGTM! Thank you!
